src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/XmlDsigSep2000.cs

```diff
@@ -0,0 +1,215 @@
+                }
+                reader.ReadEndElement();
+                return keyIdentifier;
+            }
+
+            public override bool SupportsCore(SecurityKeyIdentifier keyIdentifier)
+            {
+                return true;
+            }
+
+            public override void WriteKeyIdentifierCore(XmlDictionaryWriter writer, SecurityKeyIdentifier keyIdentifier)
+            {
+                writer.WriteStartElement(CoreWCF.XD.XmlSignatureDictionary.Prefix.Value, LocalName, NamespaceUri);
+                bool clauseWritten = false;
+                foreach (SecurityKeyIdentifierClause clause in keyIdentifier)
+                {
+                    _securityTokenSerializer.InnerSecurityTokenSerializer.WriteKeyIdentifierClause(writer, clause);
+                    clauseWritten = true;
+                }
+                writer.WriteEndElement(); // KeyInfo
+                if (!clauseWritten)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new System.Exception("NoKeyInfoClausesToWrite"));
+                }
+            }
+        }
+
+        // so far, we only support two types of X509Data directly under KeyInfo  - X509Certificate and X509SKI
+        //   <ds:X509Data>
+        //     <ds:X509Certificate>...</ds:X509Certificate>
+        //      or
+        //     <X509SKI>... </X509SKI>
+        //   </ds:X509Data>
+        // only support 1 certificate right now
+        internal class X509CertificateClauseEntry : SecurityTokenSerializer.KeyIdentifierClauseEntry
+        {
+            protected override XmlDictionaryString LocalName
+            {
+                get
+                {
+                    return CoreWCF.XD.XmlSignatureDictionary.X509Data;
+                }
+            }
+
+            protected override XmlDictionaryString NamespaceUri
+            {
+                get
+                {
+                    return CoreWCF.XD.XmlSignatureDictionary.Namespace;
+                }
+            }
+
+            public override SecurityKeyIdentifierClause ReadKeyIdentifierClauseCore(XmlDictionaryReader reader)
+            {
+                SecurityKeyIdentifierClause ski = null;
+                reader.ReadStartElement(CoreWCF.XD.XmlSignatureDictionary.X509Data, NamespaceUri);
+                while (reader.IsStartElement())
+                {
+                    if (ski == null && reader.IsStartElement(CoreWCF.XD.XmlSignatureDictionary.X509Certificate, NamespaceUri))
+                    {
+                        X509Certificate2 certificate = null;
+                        if (!SecurityUtils.TryCreateX509CertificateFromRawData(reader.ReadElementContentAsBase64(), out certificate))
+                        {
+                            throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new System.Exception(SR.InvalidX509RawData));
```

Comment:
Exception should be SecurityMessageSerializationException

Reply:
I hadn't done because all signature are wrapping system exception with one arg.. but ported now.. (assuming usability in future)